mono/metadata/threads.c

```diff
@@ -1603,9 +1603,25 @@ ves_icall_System_Threading_InternalThread_Thread_free_internal (MonoInternalThre
 	CloseHandle (this_obj->native_handle);
 #endif
 
-	if (this_obj->synch_cs) {
-		MonoCoopMutex *synch_cs = this_obj->synch_cs;
+	gpointer old_val = this_obj->synch_cs;
+	while (TRUE) {
+		gpointer new_old_val = mono_atomic_cas_ptr ((gpointer *)&this_obj->synch_cs, NULL /* set */, old_val /* compare */);
+		if (new_old_val == old_val)
+			break;
+		else
+			old_val = new_old_val;
+	}
+
+	if (old_val) {
+		MonoCoopMutex *synch_cs = (MonoCoopMutex *) old_val;
 		this_obj->synch_cs = NULL;
```

Comment:
Why do you still need this assignment?  didn't the CAS just set `this->synch_cs` to NULL?

Reply:
You're right